### PR TITLE
Clarified APF and rbac policies wrt object namespace and name

### LINKED ISSUE #89772: Describe how resource namespace and name are determined for get/list/watch requests, in authorization and API Priority and Fairness
**What happened**:
I have a client that needs to watch just one object, and I want to write RBAC that grants it only the privilege to watch one object.  But I can not do that because a watch only operates on lists.
#89755 

**What you expected to happen**:
I expected to be able to write a client that watches only one object and grant it only the privilege to do that.

**How to reproduce it (as minimally and precisely as possible)**:

**Anything else we need to know?**:

**Environment**:
- Kubernetes version (use `kubectl version`):
- Cloud provider or hardware configuration:
- OS (e.g: `cat /etc/os-release`):
- Kernel (e.g. `uname -a`):
- Install tools:
- Network plugin and version (if this is a network-related bug):
- Others:

### PR DESCRIPTION
**Pull request by MikeSpreitzer**
**What type of PR is this?**
> Uncomment only one ` /kind <>` line, hit enter to put that in a new line, and remove leading whitespace from that line:
>
> /kind api-change
> /kind bug
> /kind cleanup
> /kind deprecation
> /kind design

/kind documentation

> /kind failing-test
> /kind feature
> /kind flake

**What this PR does / why we need it**:
Explicitly noted the different treatment of the different sources
of target namespace and target name.

Other minor revisions for correctness.

**Which issue(s) this PR fixes**:

Fixes #89772

**Special notes for your reviewer**:

**Does this PR introduce a user-facing change?**:

```release-note
NONE
```

**Additional documentation e.g., KEPs (Kubernetes Enhancement Proposals), usage docs, etc.**:

```docs

```

### CODE REVIEW

**Comment by MikeSpreitzer** [comment]
@kubernetes/sig-api-machinery-bugs

**Comment by MikeSpreitzer** [comment]
/cc @liggitt 
@lavalamp 
@deads2k 
@yue9944882

**Comment by MikeSpreitzer** [comment]
@kubernetes/sig-auth-bugs

**Comment by MikeSpreitzer** [comment]
@MikeSpreitzer

File: staging/src/k8s.io/api/rbac/v1/types.go [hunk 1]
```diff
@@ -118,6 +121,9 @@ type Role struct {
 // RoleBinding references a role, but does not contain it.  It can reference a Role in the same namespace or a ClusterRole in the global namespace.
 // It adds who information via Subjects and namespace information by which namespace it exists in.  RoleBindings in a given
 // namespace only have effect in that namespace.
+// The namespace of the request is taken fromn the request URL's path
+// if present there, otherwise is the empty string; a namespace
+// specified in a field selector is ignored.
```
Comment (liggitt) [inline staging/src/k8s.io/api/rbac/v1/types.go:126]: this feels somewhat out of place here... there are multiple ways to trigger a query to the authorization system (for example, a subjectaccessreview, or by making a regular API request)...

File: staging/src/k8s.io/api/rbac/v1/types.go [hunk 2]
```diff
@@ -57,7 +57,10 @@ type PolicyRule struct {
 	// Resources is a list of resources this rule applies to.  ResourceAll represents all resources.
 	// +optional
 	Resources []string `json:"resources,omitempty" protobuf:"bytes,3,rep,name=resources"`
-	// ResourceNames is an optional white list of names that the rule applies to.  An empty set means that everything is allowed.
+	// ResourceNames is an optional white list of object names that the rule applies to.  An empty set means that everything is allowed.
+	// The object name is taken from the request URL's path if present
+	// there, otherwise from the relevant field selectoer in the URL's
+	// query if present there, otherwise is the empty string.
```
Comment (liggitt) [inline staging/src/k8s.io/api/rbac/v1/types.go:63]: this also seems out of place... the RBAC types do not describe how all attributes are determined (verbs, API groups, resources, etc)... that is the responsibility of whatever is calling the authorization system (either the subjectaccessreview API caller, or the API server)

File: staging/src/k8s.io/api/rbac/v1/types.go (hunk 1 above)
Comment (liggitt) [reply staging/src/k8s.io/api/rbac/v1/types.go:126]: making sure https://kubernetes.io/docs/reference/access-authn-authz/authorization/ accurately reflects how attributes are determined for API requests, then referencing that and subjectaccessreview docs from here might be better

File: staging/src/k8s.io/api/flowcontrol/v1alpha1/types.go [hunk 3]
```diff
@@ -239,28 +242,32 @@ type ResourcePolicyRule struct {
...
-	// specified namespace but does not match a request that _does
-	// not specify_ a namespace (see the `clusterScope` field for
-	// that).
+	// matches.  A request that specifies a target namespace in the
+	// URL's path matches only if either (a) this list contains that
+	// target namespace or (b) this list contains "*".  Note that "*"
+	// matches any specified namespace but does not match a request
+	// that _does not specify_ a namespace in the URL's path (see the
+	// `clusterScope` field for that).  Note that a namespace
+	// specified in a field selector in the URL's query paraneter is
```
Comment (lavalamp) [inline staging/src/k8s.io/api/flowcontrol/v1alpha1/types.go:269]: parameter

Comment (lavalamp) [inline staging/src/k8s.io/api/flowcontrol/v1alpha1/types.go:269]:
```suggestion
	// specified in a field selector in the URL's query parameter
	// is just a filter, and does not cause the cluster-scoped request
	// to be treated as if it were namespace scoped.
```

File: staging/src/k8s.io/api/flowcontrol/v1alpha1/types.go [hunk 4]
```diff
@@ -223,8 +223,11 @@ type ServiceAccountSubject struct {
 // ResourcePolicyRule matches a resource request if and only if: (a)
 // at least one member of verbs matches the request, (b) at least one
 // member of apiGroups matches the request, (c) at least one member of
-// resources matches the request, and (d) least one member of
-// namespaces matches the request.
+// resources matches the request, and (d) either (d1) clusterScope is
+// true and the request does not specify a namepace (which happens
```
Comment (lavalamp) [inline staging/src/k8s.io/api/flowcontrol/v1alpha1/types.go:227]:
```suggestion
// true and the request does not (or may not) specify a namepace
```

The parenthetical makes this super hard to read, I'd omit it and rely on the field documentation, whether or not you want to include my suggested hint here :)

File: staging/src/k8s.io/api/rbac/v1/types.go (hunk 2 above)
Comment (lavalamp) [reply staging/src/k8s.io/api/rbac/v1/types.go:63]: I actually think this is a relevant thing for consumers / producers of these api objects to want to know. I'd just preface it with "If the request attributes to be matched are sourced from an API request, then..."

File: staging/src/k8s.io/api/rbac/v1/types.go (hunk 1 above)
Comment (lavalamp) [reply staging/src/k8s.io/api/rbac/v1/types.go:126]: Would this go better in the Subject type, by the namespace field? I think it's fine to explain the difference between cluster and namespaced scopes in multiple places. I think explaining how the information is extracted from an ordinary API request is instructive.

**Comment by lavalamp** [comment]
/reopen

Looks like this is waiting for fixes to the comments?

**Pull request closed without merging**